src/util/mod.rs

```diff
@@ -54,3 +55,31 @@ pub fn make_spirv<'a>(data: &'a [u8]) -> super::ShaderSource<'a> {
     );
     super::ShaderSource::SpirV(words)
 }
+
+/// Asynchronously read the contents of a buffer.
+pub fn read_buffer(device: &super::Device, queue: &super::Queue, buffer: &super::BufferSlice) -> impl Future<Output=Result<Vec<u8>, super::BufferAsyncError>> + Send {
+    let size = match buffer.size {
+        Some(size) => size.into(),
+        None => buffer.buffer.map_context.lock().total_size - buffer.offset,
+    };
+
+    let download = device.create_buffer(&super::BufferDescriptor {
+        size,
+        usage: super::BufferUsage::COPY_DST | super::BufferUsage::MAP_READ,
+        mapped_at_creation: false,
+        label: None,
+    });
+
+    let mut encoder = device.create_command_encoder(&super::CommandEncoderDescriptor { label: None });
+    encoder.copy_buffer_to_buffer(buffer.buffer, buffer.offset, &download, 0, size);
+    let command_buffer: super::CommandBuffer = encoder.finish();
+    queue.submit(Some(command_buffer));
+
+    let fut = download.slice(..).map_async(super::MapMode::Read);
+    async move {
+        fut.await?;
+        let data = download.slice(..).get_mapped_range().to_vec();
```

Comment:
Would it make sense to return something like:
```rust
struct ReadBuffer {
  buffer: crate::Buffer,
}
impl Deref for ReadBuffer{
  type Item = [u8];
  fn deref(&self) -> &[u8] {
    self.buffer.slice(..).get_mapped_range()
  }
}
```
This wouldn't make it any less usable, but would avoid an unnecessary heap allocation and the data copy for many users of this new API.

Reply:
That might work. You are thinking that unmapping the buffer would just be handled when ReadBuffer was dropped?

Reply:
yep

Reply:
technically, `wgpu` should allow dropping a mapped buffer and unmap it internally, too

Reply:
I tried some variations on your version, but I couldn't find any that have workable lifetimes. The problem was that the return value from Buffer::slice() is valid only for the current scope, but deref() needs to return a slice that references it.

Reply:
I played a bit with it, and indeed you are right. A slight change is needed in order to expose this.
See https://github.com/kvark/wgpu-rs/commit/f5ea067cdb3629454cd0c4dcd17103d6bfef7dfa

Reply:
What are you thinking on this? Do you want to pull that change into wgpu-rs and then I can rebase this PR on it?

Reply:
If you can just cherry-pick this commit into your PR, that would be fine.